composer/optim/decoupled_weight_decay.py

```diff
@@ -350,6 +349,54 @@ def step(self, closure=None):
 
         return loss
 
+    def dist_reduce_metrics(self, optimizer_metrics):
+        for metric in optimizer_metrics:
+            if metric.startswith('l2_norm'):
+                reduced = optimizer_metrics[metric]
+                if dist.get_world_size() > 1:
+                    dist.all_reduce(reduced, reduce_operation='SUM')
+
+                optimizer_metrics[metric] = math.sqrt(reduced)
+            elif metric.startswith('cosine'):
+                reduced = optimizer_metrics[metric]
+                if dist.get_world_size() > 1:
+                    dist.all_reduce(reduced, reduce_operation='SUM')
+
+                _, vectors, layer = tuple(metric.split('/'))
+
+                A, B = tuple(vectors.split('_'))
+
+                A_reduced_norm = optimizer_metrics[f'l2_norm/{A}/{layer}']
+                B_reduced_norm = optimizer_metrics[f'l2_norm/{B}/{layer}']
+                optimizer_metrics[metric] = reduced / (A_reduced_norm * B_reduced_norm)
+            else:
+                reduced = optimizer_metrics[metric]
+                if dist.get_world_size() > 1:
+                    dist.all_reduce(reduced, reduce_operation='SUM')
+                optimizer_metrics[metric] = reduced / dist.get_world_size()
+
+        return optimizer_metrics
+
+    def pre_reduce_metrics(self, optimizer_metrics):
+        # Some metrics require preprocessing to reduce across ranks correctly
+
+        for metric in optimizer_metrics:
+            if metric.startswith('l2_norm'):
+                # L2 norms need to be squared, before they are reduced via summation
+                optimizer_metrics[metric] = optimizer_metrics[metric]**2
+            elif metric.startswith('cosine'):
+                _, vectors, layer = tuple(metric.split('/'))
+
+                A, B = tuple(vectors.split('_'))
+
+                # L2 norm would've been squared in previous branch
+                A_rank_subset_norm = math.sqrt(optimizer_metrics[f'l2_norm/{A}/{layer}'])
+                B_rank_subset_norm = math.sqrt(optimizer_metrics[f'l2_norm/{B}/{layer}'])
+
+                optimizer_metrics[metric] *= A_rank_subset_norm * B_rank_subset_norm
```

Comment:
I don't think this works -- you assume L2 norms are processed and squared before cosine norms

Reply:
They should be iterated in order of the key names in the class' metric_functions!

Reply:
Hm... that's a little hacky... I would prefer to explicitly sort it. Can we do something like:

```
# Sort L2 norms first so they are squared before other metrics, which depend on squared values
metrics = optimizer_metrics.keys()
metrics = sorted(metrics, key=lambda metric: 0 if 'l2_norm' in metric else 1)  
```

and then put a comment here that says "L2 norms have been already processed and squared"